mir_build: Classify `TestableCase::Constant` into multiple sub-kinds

Zalathar opened the pull request:
In match lowering, when choosing a test for a `TestableCase::Constant`, there is some ad-hoc logic for inspecting the pattern type and deciding what kind of test is suitable. There is also some very similar logic later, when partitioning cases into buckets based on the chosen test.

Instead of having that ad-hoc logic in multiple places, I think it's better to perform an up-front classification when lowering `thir::PatKind::Constant` to `TestableCase::Constant`, and then have the later steps simply match on an enum variant.

There should be no change to the resulting built MIR.

(I will note that the logic/invariants involved are a bit unclear, so there is a risk of accidental minor differences.)

Zalathar committed:
Classify `TestableCase::Constant` into multiple sub-kinds

Nadrieril commented:
r? me

Zalathar commented on compiler/rustc_mir_build/src/builder/matches/mod.rs:
```diff
@@ -1276,6 +1276,25 @@ impl<'tcx> TestableCase<'tcx> {
+    /// - floating-point primitives, e.g. `f32`, `f64`
+    /// - raw pointers derived from integer values
+    /// - pattern types, e.g. `pattern_type!(u32 is 1..)`
+    Eq,
```
Currently, range tests can only be initiated by range patterns. Those range tests will still put suitable constants into the “failure” bucket regardless of `PatConstKind`.

Zalathar replied:
Pattern types are a bit confusing. I believe the current behaviour (before this PR) is to *not* treat them as int-or-char, and instead use special logic in the equality test to transmute the scrutinee value from pattern type to underlying type.

Zalathar replied:
I agree that the `Eq` kind is a bit confusing, but that's mostly a consequence of this PR trying to mostly reflect the existing behaviour, which is already confusing.

E.g. The reason the Range/Constant case uses `kind: _` is because that *is* the pre-existing behaviour.

I have a local follow-up draft that tries to split `Eq` into “string” and ”scalar”, where the scalar case covers floats, raw pointers (derived from integers), and pattern types (with their special extra transmute step). But that involves some further changes to split apart the `TestKind::Eq` lowering into two separate code paths.

Zalathar replied:
> Huh, weird. Does that mean they don't get a `SwitchInt`, and don't support range patterns?

Yeah, I believe they currently don't get a `SwitchInt`, and probably don't support range patterns.

From what I can tell, matching for pattern-types is still in a very underbaked state. It works well enough to pass tests, but hasn't yet been nicely integrated into the rest of the match-lowering design.

Nadrieril commented:
> (I'm looking into #project-deref-patterns threads to get some context for why deref-pattern types are set up the way they currently are.)

I think it's this thread: [#project-deref-patterns > The string literal problem](https://rust-lang.zulipchat.com/#narrow/channel/281601-project-deref-patterns/topic/The.20string.20literal.20problem/with/508252214).

Zalathar commented:
I didn't mention this explicitly earlier, but my philosophy with this PR was to try to reflect the existing behaviour as much as possible. That produces results that are a bit weird/confusing, because the current behaviour is already a bit weird/confusing.

I plan to do some follow-up work to try to improve things further, so this PR represents an intermediate stepping-stone, not a final result.

Nadrieril commented on compiler/rustc_mir_build/src/builder/matches/buckets.rs:
```diff
@@ -291,7 +289,7 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 }
             }
-            (TestKind::Range(range), &TestableCase::Constant { value }) => {
+            (TestKind::Range(range), &TestableCase::Constant { value, kind: _ }) => {
```
I'd be happy to merge if you add `PatConstKind::Float` and change this line:
```suggestion
            (TestKind::Range(range), &TestableCase::Constant { value, kind: Bool | IntOrChar | Float }) => {
```
Otherwise I feel like this is slightly adding confusion on top of the existing one.

Zalathar committed:
Split out a separate `PatConstKind::Float`

Unlike the other types covered by `PatConstKind::Other`, const-float patterns
can also interact with range patterns.

Zalathar commented:
New changes:
- Renamed `PatConstKind::Eq` to `Other`
- Added a second patch that split out a separate `PatConstKind::Float`, and replaces a lot of the `kind: _` matches with more specific matches
